### PR TITLE
Python challenges

### PR DESCRIPTION
**Pull request by CheezItMan**
@jbieniosek Here are a couple of Python Challenges for Functions.  Let me know if this is what you had in mind, it's a good intro to making a python challenge.

### CODE REVIEW

File: 00-preparing-for-ada/functions.md [hunk 1]
```diff
@@ -143,22 +143,253 @@ Next, we're going to follow the same pattern as `guess_the_number`.  We will sta
...
+    @patch('sys.stdout', new_callable=io.StringIO)
+    def test_p_two(self, mock_stdout):
+        # Arrange
+        input_letter = "z"
+        original_input = mock.builtins.input
+        mock.builtins.input = lambda _: input_letter
+        answer = p.letter_in_snowman_word()
+
+        assert re.match('Letter not found', mock_stdout.getvalue())
+        assert answer == False
```
Comment (kaidamasaki) [inline 00-preparing-for-ada/functions.md:356]: If we just want to make sure something is falsey I'd recommend:

```suggestion
        assert not answer
```

If you want to make sure that they returned a literal `False` then I'd recommend:

```suggestion
        assert answer is False
```

File: 00-preparing-for-ada/functions.md [hunk 2]
```diff
@@ -143,22 +143,253 @@ Next, we're going to follow the same pattern as `guess_the_number`.  We will sta
...
+def eprint(*args, **kwargs):
+    print(*args, file=sys.stderr, **kwargs)
+
+class TestPython1(unittest.TestCase):
+
+    @patch('sys.stdout', new_callable=io.StringIO)
+    def test_p_one(self, mock_stdout):
+        # Arrange
+        input_letter = "a"
+        original_input = mock.builtins.input
```
[the remark this answers is not in the document]
Comment (CheezItMan) [reply 00-preparing-for-ada/functions.md]: Updated, I'd tried that earlier, and failed for some reason and fell back on this style.  

Now the mocking works.  Enough sleep helps.

File: 00-preparing-for-ada/functions.md [hunk 3]
```diff
@@ -151,25 +151,245 @@ Next, we're going to follow the same pattern as `guess_the_number`.  We will sta
...
+
+def get_letter_from_user():
+    letter = input("Please enter a letter > ")
+    if (len(letter) > 1 or not letter.isalpha()):
+        print("Invalid letter please enter a single character.")
+    
+    return letter
+
+
+def letter_in_snowman_word():
```
Comment (jbieniosek) [inline 00-preparing-for-ada/functions.md:312]:
```suggestion
def snowman():
```

File: 00-preparing-for-ada/functions.md [hunk 4]
```diff
@@ -151,25 +151,245 @@ Next, we're going to follow the same pattern as `guess_the_number`.  We will sta
...
+
+class TestPython2(unittest.TestCase):
+
+    @patch('sys.stdout', new_callable=io.StringIO)
+    def test_p_one(self, mock_stdout):
+        # Arrange
+        input_letter = "a"
+        with unittest.mock.patch('builtins.input', return_value=input_letter):
+            # Act
+            answer = p.letter_in_snowman_word()
```
Comment (jbieniosek) [inline 00-preparing-for-ada/functions.md:338]:
```suggestion
            answer = p.snowman()
```

File: 00-preparing-for-ada/functions.md [hunk 5]
```diff
@@ -151,25 +151,245 @@ Next, we're going to follow the same pattern as `guess_the_number`.  We will sta
...
+        assert answer # True if it's in the snowman word
+
+
+    @patch('sys.stdout', new_callable=io.StringIO)
+    def test_p_two(self, mock_stdout):
+        # Arrange
+        input_letter = "z"
+        with unittest.mock.patch('builtins.input', return_value=input_letter):
+            # Act
+            answer = p.letter_in_snowman_word()
```
Comment (jbieniosek) [inline 00-preparing-for-ada/functions.md:351]:
```suggestion
            answer = p.snowman()
```

File: 00-preparing-for-ada/functions.md [hunk 6]
```diff
@@ -151,25 +151,245 @@ Next, we're going to follow the same pattern as `guess_the_number`.  We will sta
...
+        assert re.match('Letter not found', mock_stdout.getvalue(), flags=re.IGNORECASE)
+        assert answer == False
+    
+    @patch('sys.stdout', new_callable=io.StringIO)
+    def test_p_three(self, mock_stdout):
+        # Arrange
+        input_letter = "p"
+        with unittest.mock.patch('builtins.input', return_value=input_letter):
+            # Act
+            answer = p.letter_in_snowman_word()
```
Comment (jbieniosek) [inline 00-preparing-for-ada/functions.md:363]:
```suggestion
            answer = p.snowman()
```

File: 00-preparing-for-ada/functions.md [hunk 7]
```diff
@@ -151,25 +151,245 @@ Next, we're going to follow the same pattern as `guess_the_number`.  We will sta
...
+        assert re.match('Letter found', mock_stdout.getvalue(), flags=re.IGNORECASE)
+        assert answer # True if it's in the snowman word
+
+    @patch('sys.stdout', new_callable=io.StringIO)
+    def test_p_four(self, mock_stdout):
+        # Arrange
+        input_letter = "e"
+        with unittest.mock.patch('builtins.input', return_value=input_letter):
+            # Act
+            answer = p.letter_in_snowman_word()
```
Comment (jbieniosek) [inline 00-preparing-for-ada/functions.md:375]:
```suggestion
            answer = p.snowman()
```

File: 00-preparing-for-ada/functions.md [hunk 8]
```diff
@@ -151,25 +151,245 @@ Next, we're going to follow the same pattern as `guess_the_number`.  We will sta
...
 
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+<!-- Replace everything in square brackets [] and remove brackets  -->
+
+### !challenge
+
+* type: code-snippet
+* language: python3.6
+* id: 2604b524-92e8-43fb-adc3-de4b29f0b451
+* title: letter_in_snowman_word
```
Comment (jbieniosek) [inline 00-preparing-for-ada/functions.md:282]:
```suggestion
* title: snowman
```

**Review by jbieniosek — CHANGES_REQUESTED** [review]

File: 00-preparing-for-ada/functions.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (CheezItMan) [reply 00-preparing-for-ada/functions.md]: Great, I had been wondering if that was a helper function that the snowman app would be using.

**Pull request merged**